File: src/types/gitlab.ts
export interface User {
  id: number;
  username: string;
  name: string;
  avatar_url?: string;
  web_url?: string;
}

export interface Milestone {
  id: number;
  iid: number;
  title: string;
  description?: string;
  state: 'active' | 'closed';
  due_date?: string;
  web_url?: string;
}

export interface TimeStats {
  time_estimate: number;
  total_time_spent: number;
  human_time_estimate?: string;
  human_total_time_spent?: string;
}

export interface PipelineStatus {
  id: number;
  status: 'pending' | 'running' | 'success' | 'failed' | 'canceled' | 'skipped';
  ref: string;
  sha: string;
  web_url: string;
}

export interface MergeRequest {
  id: number;
  iid: number;
  title: string;
  description: string;
  state: 'opened' | 'merged' | 'closed';
  source_branch: string;
  target_branch: string;
  author: User;
  assignees: User[];
  reviewers: User[];
  upvotes: number;
  downvotes: number;
  pipeline?: PipelineStatus;
  has_conflicts: boolean;
  draft: boolean;
  labels: string[];
  created_at: string;
  updated_at: string;
  merged_at?: string;
  closed_at?: string;
  web_url: string;
  project_id: number;
}

export interface GitLabIssue {
  id: number;
  iid: number;
  title: string;
  description: string;
  state: 'opened' | 'closed';
  labels: string[];
  assignees: User[];
  author: User;
  milestone?: Milestone;
  due_date?: string;
  weight?: number;
  time_stats: TimeStats;
  created_at: string;
  updated_at: string;
  closed_at?: string;
  web_url: string;
  project_id: number;
}

export interface Job {
  id: number;
  name: string;
  status: 'pending' | 'running' | 'success' | 'failed' | 'canceled' | 'skipped';
  stage: string;
  created_at: string;
  started_at?: string;
  finished_at?: string;
  duration?: number;
  web_url: string;
}

export interface Pipeline {
  id: number;
  iid: number;
  status: 'pending' | 'running' | 'success' | 'failed' | 'canceled' | 'skipped';
  ref: string;
  sha: string;
  web_url: string;
  created_at: string;
  updated_at: string;
  started_at?: string;
  finished_at?: string;
  duration?: number;
  jobs: Job[];
  user: User;
  project_id: number;
}

export interface MRFilters {
  state?: 'opened' | 'merged' | 'closed' | 'all';
  author_id?: number;
  assignee_id?: number;
  labels?: string[];
  target_branch?: string;
}

export interface IssueFilters {
  state?: 'opened' | 'closed' | 'all';
  author_id?: number;
  assignee_id?: number;
  labels?: string[];
  milestone?: string;
}

export interface PipelineFilters {
  status?: 'pending' | 'running' | 'success' | 'failed' | 'canceled' | 'skipped';
  ref?: string;
  username?: string;
}

export interface IssueUpdate {
  title?: string;
  description?: string;
  state_event?: 'close' | 'reopen';
  assignee_ids?: number[];
  labels?: string[];
  due_date?: string;
}

File: src-tauri/src/models/gitlab.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct User {
    pub id: i64,
    pub username: String,
    pub name: String,
    pub avatar_url: Option<String>,
    pub web_url: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Milestone {
    pub id: i64,
    pub iid: i64,
    pub title: String,
    pub description: Option<String>,
    pub state: String,
    pub due_date: Option<String>,
    pub web_url: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TimeStats {
    pub time_estimate: i64,
    pub total_time_spent: i64,
    pub human_time_estimate: Option<String>,
    pub human_total_time_spent: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PipelineStatus {
    pub id: i64,
    pub status: String,
    #[serde(rename = "ref")]
    pub ref_name: String,
    pub sha: String,
    pub web_url: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct MergeRequest {
    pub id: i64,
    pub iid: i64,
    pub title: String,
    pub description: String,
    pub state: String,
    pub source_branch: String,
    pub target_branch: String,
    pub author: User,
    pub assignees: Vec<User>,
    pub reviewers: Option<Vec<User>>,
    pub upvotes: i32,
    pub downvotes: i32,
    pub pipeline: Option<PipelineStatus>,
    pub has_conflicts: bool,
    pub draft: bool,
    pub labels: Vec<String>,
    pub created_at: String,
    pub updated_at: String,
    pub merged_at: Option<String>,
    pub closed_at: Option<String>,
    pub web_url: String,
    pub project_id: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GitLabIssue {
    pub id: i64,
    pub iid: i64,
    pub title: String,
    pub description: String,
    pub state: String,
    pub labels: Vec<String>,
    pub assignees: Vec<User>,
    pub author: User,
    pub milestone: Option<Milestone>,
    pub due_date: Option<String>,
    pub weight: Option<i32>,
    pub time_stats: TimeStats,
    pub created_at: String,
    pub updated_at: String,
    pub closed_at: Option<String>,
    pub web_url: String,
    pub project_id: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Job {
    pub id: i64,
    pub name: String,
    pub status: String,
    pub stage: String,
    pub created_at: String,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub duration: Option<f64>,
    pub web_url: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Pipeline {
    pub id: i64,
    pub iid: i64,
    pub status: String,
    #[serde(rename = "ref")]
    pub ref_name: String,
    pub sha: String,
    pub web_url: String,
    pub created_at: String,
    pub updated_at: String,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub duration: Option<f64>,
    pub user: Option<User>,
    pub project_id: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PipelineWithJobs {
    #[serde(flatten)]
    pub pipeline: Pipeline,
    pub jobs: Vec<Job>,
}
